File: examples/controlBootstrap/teardown.mjs
// Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
// Caracal, a product of Garudex Labs
//
// Removes the agent environment in reverse dependency order: policy, resource, provider, then the agent application.

import { clientFromEnv, SCOPES, AGENT, PROVIDER, RESOURCE, POLICY, byIdentifier, byName } from './plan.mjs'

export async function teardown(client, log = console.log) {
  const removed = []
  const policy = byName(await client.invoke('policy', 'list'), POLICY.name)
  if (policy) {
    await client.invoke('policy', 'delete', { id: policy.id })
    removed.push(`policy:${policy.id}`)
    log(`- policy ${POLICY.name} deleted`)
  }
  const resource = byIdentifier(await client.invoke('resource', 'list'), RESOURCE.identifier)
  if (resource) {
    await client.invoke('resource', 'delete', { id: resource.id })
    removed.push(`resource:${resource.id}`)
    log(`- resource ${RESOURCE.identifier} deleted`)
  }
  const provider = byIdentifier(await client.invoke('identity-provider', 'list'), PROVIDER.identifier)
  if (provider) {
    await client.invoke('identity-provider', 'delete', { id: provider.id })
    removed.push(`provider:${provider.id}`)
    log(`- identity-provider ${PROVIDER.identifier} deleted`)
  }
  const agent = byName(await client.invoke('app', 'list'), AGENT.name)
  if (agent) {
    await client.invoke('app', 'delete', { id: agent.id })
    removed.push(`app:${agent.id}`)
    log(`- app ${AGENT.name} deleted`)
  }
  return removed
}

async function main() {
  const client = clientFromEnv(SCOPES.teardown)
  const removed = await teardown(client)
  console.log(`teardown complete (${removed.length} object(s) removed)`)
}

if (import.meta.url === `file://${process.argv[1]}`) {
  main().catch((err) => {
    console.error(err instanceof Error ? err.message : String(err))
    process.exitCode = 1
  })
}

File: examples/lynxCapital/scripts/teardown.py
"""
Copyright (C) 2026 Garudex Labs.  All Rights Reserved.
Caracal, a product of Garudex Labs

Removes the Lynx Capital multi-tenant deployment: per-tenant DCR applications, the policy
set, policies, resources, and the managed platform application.
"""
from __future__ import annotations

import sys
from pathlib import Path

sys.path.insert(0, str(Path(__file__).resolve().parent.parent))

from app import tenancy
from control_client import ControlClient, config_from_env, find_by_identifier, find_by_name


def remove(client: ControlClient, command: str, item: dict | None, label: str) -> None:
    if not item or not item.get("id"):
        return
    client.invoke(command, "delete", {"id": item["id"]})
    print(f"{label} removed: {item.get('identifier') or item.get('name')}")


def main() -> None:
    model = tenancy.load_model()
    client = ControlClient(config_from_env())

    app_list = client.invoke("app", "list")
    for tenant in model.tenants:
        remove(client, "app", find_by_name(app_list, tenant.dcrApplicationName), "tenant DCR app")

    remove(client, "policy-set", find_by_name(client.invoke("policy-set", "list"), model.policySet.name), "policy-set")

    policy_list = client.invoke("policy", "list")
    for name, _ in tenancy.policy_files():
        remove(client, "policy", find_by_name(policy_list, name), "policy")

    resource_list = client.invoke("resource", "list")
    for resource in model.resources:
        remove(client, "resource", find_by_identifier(resource_list, resource.identifier), "resource")

    remove(client, "app", find_by_name(client.invoke("app", "list"), model.platform.applicationName), "managed app")
    print(f"removed platform + {len(model.tenants)} tenants")


if __name__ == "__main__":
    main()
